client/src/main/java/io/pravega/client/stream/impl/Pinger.java

```diff
@@ -99,32 +100,34 @@ private void startPeriodicPingTxn() {
      */
     private void pingTransactions() {
         log.info("Start sending transaction pings.");
-
-        List<UUID> stopPingTxns = new ArrayList<>();
-        completedTxns.drainTo(stopPingTxns);
-        txnList.removeAll(stopPingTxns); // remove completed transactions from the pingable transaction list.
-        txnList.forEach(uuid -> {
-            try {
-                log.debug("Sending ping request for txn ID: {} with lease: {}", uuid, txnLeaseMillis);
-                controller.pingTransaction(stream, uuid, txnLeaseMillis)
-                          .whenComplete((status, e) -> {
-                              if (e != null) {
-                                  log.warn("Ping Transaction for txn ID:{} failed", uuid, unwrap(e));
-                              } else if (Transaction.PingStatus.ABORTED.equals(status) || Transaction.PingStatus.COMMITTED.equals(status)) {
-                                  completedTxns.offer(uuid);
-                              }
-                          });
-            } catch (Exception e) {
-                // Suppressing exception to prevent future pings from not being executed. 
-                log.warn("Encountered exception when attepting to ping transactions", e);
-            }
-        });
+        synchronized (lock) {
+            txnList.removeAll(completedTxns);  // remove completed transactions from the pingable transaction list.
+            completedTxns.clear();
+            txnList.forEach(uuid -> {
+                try {
+                    log.debug("Sending ping request for txn ID: {} with lease: {}", uuid, txnLeaseMillis);
+                    controller.pingTransaction(stream, uuid, txnLeaseMillis)
+                              .whenComplete((status, e) -> {
+                                  if (e != null) {
+                                      log.warn("Ping Transaction for txn ID:{} failed", uuid, unwrap(e));
+                                  } else if (Transaction.PingStatus.ABORTED.equals(status) || Transaction.PingStatus.COMMITTED.equals(status)) {
+                                      completedTxns.add(uuid);
+                                  }
+                              });
+                } catch (Exception e) {
+                    // Suppressing exception to prevent future pings from not being executed.
+                    log.warn("Encountered exception when attempting to ping transactions", e);
+                }
+            });
+        }
         log.trace("Completed sending transaction pings.");
     }
 
     @Override
     public void close() {
         log.info("Closing Pinger periodic task");
-        ExecutorServiceHelpers.shutdown(executor);
+        if (scheduledFuture != null) {
```

Comment:
This is not thread safe. In between line 129 and 130, someone could have executed some code to set this to null.

If you change this to `AtomicReference`, then you can do:
```
val toCancel = scheduledFuture.getAndSet(null);
if(toCancel != null){
     toCancel.cancel(true);
}
```

Reply:
done, used `AtomicReference`.